plan/match_property.go

```diff
@@ -20,73 +20,89 @@ import (
 )
 
 // matchProperty implements PhysicalPlan matchProperty interface.
-func (ts *PhysicalTableScan) matchProperty(prop requiredProperty, rowCounts []uint64, _ ...*physicalPlanInfo) *physicalPlanInfo {
-	rowCount := float64(rowCounts[0])
+func (ts *PhysicalTableScan) matchProperty(prop *requiredProperty, infos ...*physicalPlanInfo) *physicalPlanInfo {
+	rowCount := float64(infos[0].count)
 	cost := rowCount * netWorkFactor
-	if len(prop) == 0 {
-		return &physicalPlanInfo{p: ts, cost: cost}
+	if len(prop.props) == 0 {
+		return &physicalPlanInfo{p: ts, cost: cost, count: infos[0].count}
 	}
-	if len(prop) == 1 && ts.pkCol != nil && ts.pkCol == prop[0].col {
+	if len(prop.props) == 1 && ts.pkCol != nil && ts.pkCol == prop.props[0].col {
 		sortedTs := *ts
-		sortedTs.Desc = prop[0].desc
+		sortedTs.Desc = prop.props[0].desc
 		sortedTs.KeepOrder = true
-		return &physicalPlanInfo{p: &sortedTs, cost: cost}
+		return &physicalPlanInfo{p: &sortedTs, cost: cost, count: infos[0].count}
 	}
-	return &physicalPlanInfo{p: ts, cost: math.MaxFloat64}
+	return &physicalPlanInfo{p: ts, cost: math.MaxFloat64, count: infos[0].count}
 }
 
 // matchProperty implements PhysicalPlan matchProperty interface.
-func (is *PhysicalIndexScan) matchProperty(prop requiredProperty, rowCounts []uint64, _ ...*physicalPlanInfo) *physicalPlanInfo {
-	rowCount := float64(rowCounts[0])
-	// currently index read from kv 2 times.
+func (is *PhysicalIndexScan) matchProperty(prop *requiredProperty, infos ...*physicalPlanInfo) *physicalPlanInfo {
+	rowCount := float64(infos[0].count)
 	cost := rowCount * netWorkFactor
 	if is.DoubleRead {
 		cost *= 2
 	}
-	if len(prop) == 0 {
-		return &physicalPlanInfo{p: is, cost: cost}
+	if len(prop.props) == 0 {
+		return &physicalPlanInfo{p: is, cost: cost, count: infos[0].count}
 	}
-	matched := 0
+	matchedIdx := 0
 	allDesc, allAsc := true, true
-	for i, indexCol := range is.Index.Columns {
-		if indexCol.Length != types.UnspecifiedLength {
+	matchedList := make([]bool, len(prop.props))
+	for i, idxCol := range is.Index.Columns {
+		if idxCol.Length != types.UnspecifiedLength {
 			break
 		}
-		if prop[matched].col.ColName.L != indexCol.Name.L {
-			if matched == 0 && i < is.accessEqualCount {
-				continue
+		findMatched := false
+		if matchedIdx < prop.sortKeyLen {
+			propCol := prop.props[matchedIdx]
+			if idxCol.Name.L == propCol.col.ColName.L {
+				findMatched = true
+				matchedList[matchedIdx] = true
+				if propCol.desc {
+					allAsc = false
+				} else {
+					allDesc = false
+				}
+				matchedIdx++
 			}
-			break
-		}
-		if prop[matched].desc {
-			allAsc = false
 		} else {
-			allDesc = false
+			for j, propCol := range prop.props {
+				if idxCol.Name.L == propCol.col.ColName.L {
+					matchedList[j] = true
+					findMatched = true
+					break
+				}
+			}
 		}
-		matched++
-		if matched == len(prop) {
+		if !findMatched && i >= is.accessEqualCount {
```

Comment:
because the required property ordered by c can match index(a,b,c) with condition a = 1 and b =1 and c < 1.